### PR TITLE
Stricter tests for grade-school

### LINKED ISSUE #660: grade-school: student name pointer assignment
Hello!

Thank you for the brilliant work you are doing!

---

After finishing this exercise and reading some of the community solutions, I saw that some of the solutions passed even though, in my opinion, they are technically not correct.

Student structure:
```
#define MAX_NAME_LENGTH 20

typedef struct {
   uint8_t grade;
   char *name;
} student_t;
```
The prototype of `add_student()`:

```
bool		add_student(const char* name ,uint8_t grade);
```
`name` from `student_t` struct is a pointer, and assigning the pointer to the `name` parameter passed to the function pass the tests.

Let's say we have a roster variable:

```
static roster_t roster;
```

If in the `add_student()` function we do:

```
roster.students[i].name = name;
```

Both pointers will point to the same address. The tests will pass because the parameter is the `expected->students[i].name`.

---

Thank you for your time!

### PR DESCRIPTION
**Pull request by ryanplusplus**
Fixes #660

### CODE REVIEW

**Review by sylenth1cs50 — APPROVED** [review]
Overall looks great. Thank you!

The only suggestion I have is to consider `snprintf`, even though `strcpy` is perfectly fine now because the length of the names is lower than `MAX_NAME_LENGTH`.

File: exercises/practice/grade-school/.meta/example.h
Comment (ryanplusplus) [reply exercises/practice/grade-school/.meta/example.h:13]: Good catch! Fixed.

File: exercises/practice/grade-school/.meta/example.c
Comment (ryanplusplus) [reply exercises/practice/grade-school/.meta/example.c:31]: :+1:

File: exercises/practice/grade-school/.meta/example.c
Comment (ryanplusplus) [reply exercises/practice/grade-school/.meta/example.c]: Good call, fixed.

**Review by sylenth1cs50 — APPROVED** [review]
Thank you very much, @ryanplusplus!

---

Everything looks good to me.

I appreciate what you and the other maintainers are doing.

**Comment by wolf99** [comment]
Think this was purposely done this way so as to make the exercise more about ordering than messing around with memory so as to be able to copy strings around the place.
Having said that this does implementation does not seem like much extra burden on the student.

File: exercises/practice/grade-school/test_grade_school.c
Comment (wolf99) [inline exercises/practice/grade-school/test_grade_school.c:18]: `MAX_NAME_LENGTH` is part of the implementation, technically students are free to change it or remove it completely... this would then break.

File: exercises/practice/grade-school/.meta/example.c
Comment (wolf99) [inline exercises/practice/grade-school/.meta/example.c:31]: If we consider the string terminator to be part of the name length, then this can chane from `<=` to `<` and the `+1`'s won't be needed when declaring the `name[]`  string in the header file.

```suggestion
   if (roster.count < MAX_STUDENTS && strlen(name) < MAX_NAME_LENGTH) {
```

File: exercises/practice/grade-school/test_grade_school.c
Comment (ryanplusplus) [reply exercises/practice/grade-school/test_grade_school.c:18]: If it's changed but not removed, this will change accordingly because it's using the macro from the implementation's header.

File: exercises/practice/grade-school/.meta/example.c
Comment (ryanplusplus) [reply exercises/practice/grade-school/.meta/example.c]: Good suggestion, I made these changes.

**Comment by ryanplusplus** [comment]
> IIRC this was exercise was purposely implemented this way so as to make the exercise focus more on sorting the pointers based on the contents (e.g. `qsort()` and `strncmp()`) rather than copying strings around the place for the sake of it and the memory management that can creep in to student solutions to allow that. Part of our previous effort to have more logic and less string manipulation in the track.
> 
> Having said that this implementation does not seem like much extra burden on the student.

Your call, I can go either way.

File: exercises/practice/grade-school/grade_school.h
Comment (wolf99) [inline exercises/practice/grade-school/grade_school.h:12]: Missed one 😉 

```suggestion
   char name[MAX_NAME_LENGTH];
```

Comment (ryanplusplus) [reply exercises/practice/grade-school/grade_school.h]: Whoops, thanks

File: exercises/practice/grade-school/test_grade_school.c
Comment (wolf99) [reply exercises/practice/grade-school/test_grade_school.c]: > Your call, I can go either way. If we decide to keep the existing behavior, we should remote `MAX_NAME_LENGTH` since it is inconsistent with using statically allocated strings.

Go with this new approach, if we get feedback that possible solutions are too complex then we can adjust again a needed.

**Comment by wolf99** [comment]
>Your call, I can go either way. If we decide to keep the existing behavior, we should remote MAX_NAME_LENGTH since it is inconsistent with using statically allocated strings.

Go with this new approach, if we get feedback that possible solutions are too complex then we can adjust again a needed.

**Review by wolf99 — APPROVED** [review]

**Pull request merged**
